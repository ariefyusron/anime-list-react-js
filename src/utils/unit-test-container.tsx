import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import themes from "@themes";

const queryClient = new QueryClient();

const renderWithTheme = (ui: ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
    </ThemeProvider>
  );
};

export { renderWithTheme };
