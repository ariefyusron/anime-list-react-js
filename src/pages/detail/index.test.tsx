import { screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";

import { renderWithTheme } from "../../utils/unit-test-container";

import DetailPage from ".";

const data = {
  data: {
    title: "Title",
    type: "Type",
    episodes: 12,
    status: "Status",
    aired: { string: "Aired" },
    broadcast: { string: "Broadcast" },
    producers: [{ name: "Producers" }, { name: "Producers2" }],
    licensors: [{ name: "Licensors" }, { name: "Licensors2" }],
    studios: [{ name: "Studios" }, { name: "Studios2" }],
    source: "Source",
    genres: [{ name: "Genres" }, { name: "Genres2" }],
    themes: [{ name: "Themes" }, { name: "Themes2" }],
    demographics: [{ name: "Demographics" }, { name: "Demographics2" }],
    duration: "Duration",
    rating: "Rating",
    score: 12,
    rank: 12,
    popularity: 12,
    members: 12,
    favorites: 12,
    images: {
      webp: {
        large_image_url: "image_url.large"
      }
    },
    synopsis: "ini panjang"
  }
};

let mockAnimeDetail = {
  data,
  isLoading: false
};

const mockNavigate = vi.fn();

describe("Detail Page", () => {
  vi.mock("react-router", async () => {
    const actual = await vi.importActual("react-router");

    return {
      ...actual,
      useNavigate: vi.fn(() => mockNavigate)
    };
  });

  beforeEach(() => {
    vi.mock("../../hooks/useAnime", () => ({
      useGetAnimeDetail: vi.fn(() => mockAnimeDetail)
    }));

    mockNavigate.mockClear();
  });

  it("render the Detail Page Loading Snapshot", () => {
    mockAnimeDetail = {
      data,
      isLoading: true
    };
    const component = renderWithTheme(<DetailPage />);

    expect(component).toMatchSnapshot();
  });

  it("render the Detail Page With Data Snapshot", () => {
    mockAnimeDetail = {
      data,
      isLoading: false
    };
    const component = renderWithTheme(<DetailPage />);

    expect(component).toMatchSnapshot();
  });

  it("should render the Detail Page Loading", () => {
    mockAnimeDetail = {
      data,
      isLoading: true
    };

    renderWithTheme(<DetailPage />);

    const buttonBack = screen.getByText("Back");
    expect(buttonBack).toBeInTheDocument();
    fireEvent.click(buttonBack);
    expect(mockNavigate).toHaveBeenCalledWith(-1);

    expect(screen.getByText("Loading...")).toBeInTheDocument();
    expect(screen.getByTestId("component-loader")).toBeInTheDocument();
  });

  it("should render the Detail Page With Data", () => {
    mockAnimeDetail = {
      data,
      isLoading: false
    };

    renderWithTheme(<DetailPage />);

    expect(screen.getByText("Title")).toBeInTheDocument();

    const imageElement = screen.getByAltText("image-detail");
    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute("src", "image_url.large");

    expect(screen.getByText("Informasi")).toBeInTheDocument();
    expect(screen.getByText("Type: Type")).toBeInTheDocument();
    expect(screen.getByText("Episodes: 12")).toBeInTheDocument();
    expect(screen.getByText("Status: Status")).toBeInTheDocument();
    expect(screen.getByText("Aired: Aired")).toBeInTheDocument();
    expect(screen.getByText("Broadcast: Broadcast")).toBeInTheDocument();
    expect(
      screen.getByText("Producers: Producers, Producers2")
    ).toBeInTheDocument();
    expect(
      screen.getByText("Licensors: Licensors, Licensors2")
    ).toBeInTheDocument();
    expect(screen.getByText("Studios: Studios, Studios2")).toBeInTheDocument();
    expect(screen.getByText("Source: Source")).toBeInTheDocument();
    expect(screen.getByText("Genres: Genres, Genres2")).toBeInTheDocument();
    expect(screen.getByText("Themes: Themes, Themes2")).toBeInTheDocument();
    expect(
      screen.getByText("Demographics: Demographics, Demographics2")
    ).toBeInTheDocument();
    expect(screen.getByText("Duration: Duration")).toBeInTheDocument();
    expect(screen.getByText("Rating: Rating")).toBeInTheDocument();

    expect(screen.getByText("Statistic")).toBeInTheDocument();
    expect(screen.getByText("Score: 12")).toBeInTheDocument();
    expect(screen.getByText("Ranked: 12")).toBeInTheDocument();
    expect(screen.getByText("Popularity: 12")).toBeInTheDocument();
    expect(screen.getByText("Members: 12")).toBeInTheDocument();
    expect(screen.getByText("Favorites: 12")).toBeInTheDocument();

    expect(screen.getByText("Synopsis")).toBeInTheDocument();
    expect(screen.getByText("ini panjang")).toBeInTheDocument();
  });
});
