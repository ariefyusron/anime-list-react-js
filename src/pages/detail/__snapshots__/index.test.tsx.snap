// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Detail Page > render the Detail Page Loading Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gjcoXW eXzRsk"
      >
        <div
          class="sc-fLDLck ljLhSS"
        >
          Back
        </div>
        <h1
          class="sc-dVBluf gLMRWv"
        >
          Loading...
        </h1>
      </div>
      <div
        data-testid="component-loader"
        style="display: flex; justify-content: center; margin-top: 100px;"
      >
        <div
          class="sc-FEMpB jtZqmY"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gjcoXW eXzRsk"
    >
      <div
        class="sc-fLDLck ljLhSS"
      >
        Back
      </div>
      <h1
        class="sc-dVBluf gLMRWv"
      >
        Loading...
      </h1>
    </div>
    <div
      data-testid="component-loader"
      style="display: flex; justify-content: center; margin-top: 100px;"
    >
      <div
        class="sc-FEMpB jtZqmY"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Detail Page > render the Detail Page With Data Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gjcoXW eXzRsk"
      >
        <div
          class="sc-fLDLck ljLhSS"
        >
          Back
        </div>
        <h1
          class="sc-dVBluf gLMRWv"
        >
          Title
        </h1>
      </div>
      <div
        class="sc-dycHyt kitVOT"
      >
        <div
          class="sc-gNZgCX hhIUKd"
        >
          <img
            alt="image-detail"
            class="sc-jbAkgO dSjlwp"
            src="image_url.large"
          />
          <div
            class="sc-kfeOyU JIrGb"
          />
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 8px; border-bottom: 1px solid #fff;"
          >
            Informasi
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Type: Type
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Episodes: 12
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Status: Status
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Aired: Aired
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Broadcast: Broadcast
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Producers: Producers, Producers2
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Licensors: Licensors, Licensors2
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Studios: Studios, Studios2
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Source: Source
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Genres: Genres, Genres2
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Themes: Themes, Themes2
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Demographics: Demographics, Demographics2
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Duration: Duration
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Rating: Rating
          </div>
        </div>
        <div
          class="sc-gNZgCX hhIUKd"
          style="flex: 1;"
        >
          <div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 8px; border-bottom: 1px solid #fff;"
            >
              Statistic
            </div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 4px;"
            >
              Score: 12
            </div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 4px;"
            >
              Ranked: 12
            </div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 4px;"
            >
              Popularity: 12
            </div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 4px;"
            >
              Members: 12
            </div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 4px;"
            >
              Favorites: 12
            </div>
          </div>
          <div>
            <div
              class="sc-kfeOyU JIrGb"
              style="margin-top: 8px; border-bottom: 1px solid #fff; margin-bottom: 4px;"
            >
              Synopsis
            </div>
            <div
              class="sc-kfeOyU JIrGb"
            >
              ini panjang
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gjcoXW eXzRsk"
    >
      <div
        class="sc-fLDLck ljLhSS"
      >
        Back
      </div>
      <h1
        class="sc-dVBluf gLMRWv"
      >
        Title
      </h1>
    </div>
    <div
      class="sc-dycHyt kitVOT"
    >
      <div
        class="sc-gNZgCX hhIUKd"
      >
        <img
          alt="image-detail"
          class="sc-jbAkgO dSjlwp"
          src="image_url.large"
        />
        <div
          class="sc-kfeOyU JIrGb"
        />
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 8px; border-bottom: 1px solid #fff;"
        >
          Informasi
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Type: Type
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Episodes: 12
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Status: Status
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Aired: Aired
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Broadcast: Broadcast
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Producers: Producers, Producers2
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Licensors: Licensors, Licensors2
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Studios: Studios, Studios2
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Source: Source
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Genres: Genres, Genres2
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Themes: Themes, Themes2
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Demographics: Demographics, Demographics2
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Duration: Duration
        </div>
        <div
          class="sc-kfeOyU JIrGb"
          style="margin-top: 4px;"
        >
          Rating: Rating
        </div>
      </div>
      <div
        class="sc-gNZgCX hhIUKd"
        style="flex: 1;"
      >
        <div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 8px; border-bottom: 1px solid #fff;"
          >
            Statistic
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Score: 12
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Ranked: 12
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Popularity: 12
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Members: 12
          </div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 4px;"
          >
            Favorites: 12
          </div>
        </div>
        <div>
          <div
            class="sc-kfeOyU JIrGb"
            style="margin-top: 8px; border-bottom: 1px solid #fff; margin-bottom: 4px;"
          >
            Synopsis
          </div>
          <div
            class="sc-kfeOyU JIrGb"
          >
            ini panjang
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
